### PR TITLE
[LD-139] chore: test coverage report

### PR DESCRIPTION
**Pull request by trOnk12**
# Pull-Request
## Description
- Add per pull request code coverage report generation. 

### Why?
 - There is no way to monitor the code coverage increase of decrease, in order to maintain high quality code, we want to have a tool allowing us to monitor the changes.
### What?
- Add a gradle configuration allowing to generate the report and add a task correctly configuring the coverage report settings.

## Links to related issues
(https://loudius.atlassian.net/browse/LD-139)

## Demo
image

## How to test
 - Try to add some code uncovered by unit tests
 - Push the changes, there should be a PR comment added indicating a decrease in code coverage

## Documentation
[Github action](https://github.com/marketplace/actions/jacoco-report) used to create a PR comment
[Jacoco](https://github.com/jacoco/jacoco) Library genereting the coverage report
## Checklist
- [x] Functionality is covered by unit tests ( not applicable )
- [x] Functionality is covered by integration tests ( not applicable )
- [x] I've updated PR description with relevant information
- [x] I've done self code review
- [x] I've manually tested if the code and app works

### CODE REVIEW

**Comment by Krzysiudan** [comment]
|Overall Project|72.41%|:x:|
|:-|:-|:-:|

> There is no coverage information present for the Files changed

File: .github/workflows/run-test.yml [hunk 1]
```diff
@@ -36,14 +38,30 @@ jobs:
         uses: ./.github/actions/prepare-android-env
 
       - name: Run test
-        run: ./gradlew test
```
Comment (nowakweronika) [inline .github/workflows/run-test.yml:39]: Question: Why is this removed?

**Review by nowakweronika — COMMENTED** [review]
Can you describe a little what the crucial part of code is doing?😅

File: .github/workflows/run-test.yml (hunk 1 above)
Comment (trOnk12) [reply .github/workflows/run-test.yml:39]: it is replaced by a task that I created codeCoverage

File: .github/workflows/run-test.yml [hunk 2]
```diff
@@ -36,14 +38,30 @@ jobs:
...
+          name: coverage-report
+          path: |
+            */build/reports/jacoco/codeCoverage/html/**
+
+      - name: Add coverage to PR
+        id: jacoco
+        uses: madrapps/jacoco-report@v1.6.1
+        with:
+          paths: |
+            */build/reports/jacoco/codeCoverage/codeCoverage.xml
```
Comment (trOnk12) [inline .github/workflows/run-test.yml:61]: here we pass the path to the codeCoverage report in xml format to the action

File: app/build.gradle [hunk 3]
```diff
@@ -93,6 +94,50 @@ android {
     }
 }
 
+tasks.register('codeCoverage', JacocoReport) {
+    dependsOn 'testDebugUnitTest'
```
Comment (trOnk12) [inline app/build.gradle:97]: dependsOn means that it will execute testDebugUnitTest when triggering codeCoverage and then all that is defined here

File: app/build.gradle [hunk 4]
```diff
@@ -93,6 +94,50 @@ android {
...
+    )
+
+    sourceDirectories.setFrom(fileTree(
+            dir: project.projectDir,
+            includes: [
+                    "src/main/java/**",
+                    "src/main/kotlin/**"
+            ]))
+
+    executionData.setFrom(fileTree(
```
Comment (trOnk12) [inline app/build.gradle:131]: The location of "execution data", i.e. the coverage data collected by the unit test runner when running the tests against classes instrumented by the JaCoCo agent.

File: app/build.gradle [hunk 5]
```diff
@@ -93,6 +94,50 @@ android {
...
 
+tasks.register('codeCoverage', JacocoReport) {
+    dependsOn 'testDebugUnitTest'
+
+    reports {
+        html.required.set(true)
+        xml.required.set(true)
+    }
+
+    classDirectories.setFrom(
```
Comment (trOnk12) [inline app/build.gradle:104]: : The compiled class files without instrumentation. Execution data references instructions in these classes, so they're necessary for locating the actual instructions.

File: app/build.gradle [hunk 6]
```diff
@@ -93,6 +94,50 @@ android {
...
+            fileTree(project.buildDir) {
+                include("**/tmp/kotlin-classes/debug/**")
+                exclude(
```
Comment (trOnk12) [inline app/build.gradle:107]: we can exclude some classes so that we they are not included in the code coverage report

File: app/build.gradle [hunk 7]
```diff
@@ -93,6 +94,50 @@ android {
...
+                        '**/androidx/**/*.*',
+                        '**/airbnb/**/*.*',
+                        "**/di/**/*.*",
+                        '**/*Dagger*.*',
+                        "**/*Screen*"
+                )
+            }
+    )
+
+    sourceDirectories.setFrom(fileTree(
```
Comment (trOnk12) [inline app/build.gradle:124]: the location of the original source code files. Optional, but allows the reporter to map bytecode instructions back to the original sources in order to include them into the report with coverage annotations.

**Comment by Krzysiudan** [comment]
|Overall Project|72.41%|:x:|
|:-|:-|:-:|

> There is no coverage information present for the Files changed

**Comment by Krzysiudan** [comment]
|Overall Project|72.41%|:x:|
|:-|:-|:-:|

> There is no coverage information present for the Files changed

File: .github/workflows/run-test.yml [hunk 8]
```diff
@@ -36,14 +38,30 @@ jobs:
...
+          token: ${{ secrets.PAT || secrets.GITHUB_TOKEN }}
```
Comment (jacek-marchwicki) [inline .github/workflows/run-test.yml]: **question:** why we have here: `${{ secrets.PAT || secrets.GITHUB_TOKEN }}`?

**Review by jacek-marchwicki — APPROVED** [review]
lgtm, but keep in mind that we need to continue to work on this subject to add support for merging test reports
